File: Software Design and Engineering/Original Version/ContactService.java
import java.util.ArrayList;
public class ContactService {
		public String firstName;
		public String lastName;
		public String PhoneNumber;
		public String Address; 
		public static String ID = "0";
		static ArrayList<Contact> contactList = new ArrayList<Contact>(0);
		
		public static String getNextID() {
			int temp = Integer.valueOf(ID);
			temp += 1;
			ID = Integer.toString(temp);
			return ID;
		}
		public static String getID() {
			return ID;
		}
		//add contact
		public static void addContact(String fName, String lName, String phone, String newAddress) {
			String ID = getNextID();
			Contact newContact = new Contact(ID, fName, lName, phone, newAddress);
			contactList.add(newContact);
		}
		
		//delete contact
		public static void deleteContact(String uniqueID) {
			for(int i = 0; i < contactList.size(); i++) {
				if(uniqueID.compareTo(contactList.get(i).getId()) == 0) {
					contactList.remove(i);
				}
			}
		}
		
		//update methods
		public static void updateFirstName(String uniqueID, String newFirstName) {
			for(int i = 0; i < contactList.size(); i++) {
				if(uniqueID.compareTo(contactList.get(i).getId()) == 0) {
						contactList.get(i).setFirstName(newFirstName);
				}
			}
		}
		public static void updateLastName(String uniqueID, String newLastName) {
			for(int i = 0; i < contactList.size(); i++) {
				if(uniqueID.compareTo(contactList.get(i).getId()) == 0) {
						contactList.get(i).setLastName(newLastName);
				}
			}
		}
		public static void updatePhoneNumber(String uniqueID, String newPhoneNumber) {
			for(int i = 0; i < contactList.size(); i++) {
				if(uniqueID.compareTo(contactList.get(i).getId()) == 0) {
						contactList.get(i).setPhoneNumber(newPhoneNumber);
				}
			}
		}
		public static void updateAddress(String uniqueID, String newAddress) {
			for(int i = 0; i < contactList.size(); i++) {
				if(uniqueID.compareTo(contactList.get(i).getId()) == 0) {
						contactList.get(i).setAddress(newAddress);
				}
			}
		}
}

File: Software Design and Engineering/Enhanced Version/ContactService.cpp
#include "ContactService.h"

void ContactService::addContact(const Contact& contact) {
    for (const auto& c : contactList) {
        if (c.getContactId() == contact.getContactId()) {
            throw std::invalid_argument("Contact ID must be unique");
        }
    }
    contactList.push_back(contact);
}

void ContactService::deleteContact(const std::string& contactId) {
    for (auto it = contactList.begin(); it != contactList.end(); ++it) {
        if (it->getContactId() == contactId) {
            contactList.erase(it);
            return;
        }
    }
    throw std::invalid_argument("Contact with ID " + contactId + " not found");
}

Contact* ContactService::findContactById(const std::string& contactId) {
    for (auto& contact : contactList) {
        if (contact.getContactId() == contactId) {
            return &contact;
        }
    }
    throw std::invalid_argument("Contact with ID " + contactId + " not found");
}

void ContactService::updateFirstName(const std::string& contactId, const std::string& newFirstName) {
    findContactById(contactId)->setFirstName(newFirstName);
}

void ContactService::updateLastName(const std::string& contactId, const std::string& newLastName) {
    findContactById(contactId)->setLastName(newLastName);
}

void ContactService::updatePhone(const std::string& contactId, const std::string& newPhone) {
    findContactById(contactId)->setPhone(newPhone);
}

void ContactService::updateAddress(const std::string& contactId, const std::string& newAddress) {
    findContactById(contactId)->setAddress(newAddress);
}

const std::vector<Contact>& ContactService::getContacts() const {
    return contactList;
}
